rfcs/0002-netapi-async-batch.md

```diff
@@ -0,0 +1,55 @@
+Feature Name: netapi-async-batch
+- Start Date: 2018-09-20
+- RFC PR: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Salt's HTTP API (NetApi) has incomplete support for the "batching mode" available via the commandline `-b` switch, in particular the `rest_cherrypy` implementation can either support asynchronous or batched commands, but not both at the same time.
+
+This RFC is about supporting batched commands in the existing asynchronous endpoint via additional parameters.
+
+# Motivation
+[motivation]: #motivation
+
+Salt can run commands in parallel on huge number of minions, and those commands might generate load on potentially unrelated parts of the user's infrastructure (eg. Web servers), hence some kind of rate limiting ("batching" in current Salt jargon) is desirable.
+
+Uyuni/SUSE Manager would need to leverage such mechanism through `rest_cherrypy`, in particular via asynchronous commands which is how the integration is done at this point.
+
+# Design
+[design]: #detailed-design
+
+Despite batching being implemented in [salt/cli/batch.py](https://github.com/saltstack/salt/blob/v2018.3.2/salt/cli/batch.py), the current `Batch` class is actually a shared component used by:
+  - [the command line `salt` command](https://github.com/saltstack/salt/blob/v2018.3.2/salt/cli/salt.py#L232)
+  - [the `rest_cherrypy` implementation](https://github.com/saltstack/salt/blob/v2018.3.2/salt/client/__init__.py#L571) (through `LocalClient.cmd_batch`)
+  - [the currently-deleted `rest_tornado` implementation](https://github.com/saltstack/salt/blob/2015.5/salt/netapi/rest_tornado/saltnado.py#L216) but only to resolve the list of minions. [The bulk of the logic has been reimplemented in saltnado.py](https://github.com/saltstack/salt/blob/2015.5/salt/netapi/rest_tornado/saltnado.py#L862-L884). Note that the whole batching code in `rest_tornado` [has been deleted](https://github.com/saltstack/salt/commit/3d8f3d18f6afa760c70db87cbbaaa71d877ca4d3) as a [response of a security issue](https://github.com/saltstack/salt/issues/38497) (lack of eauth support) - but in principle this code here works
+
+Please note that the `Batch` class currently works synchronously only - this is of course OK for the CLI, it's acceptable for direct `LocalClient` use by third-party Python code, it's probably also acceptable for `rest_cherrypy` (although asynchronous calls are not supported) or direct `LocalClient` but it's definitely not adequate for `rest_tornado`.
+
+We propose to move the handling of batches into `salt-master`, so that long-term there is only one implementation and all clients/NetApi clients get support "for free".
+
+Implementation ideas:
+ - a new parameter is added to the [publish method in `ClearFuncs`](https://github.com/saltstack/salt/blob/develop/salt/master.py#L2035) which is the receiving endpoint for commands from clients, or a new method is created (eg. `publish_batch`)
```

Comment:
If we're going to do this, I think it's cleaner to split it out into multiple methods and keep `publish` as it is.

Reply:
@cachedout I am second to this. Current publish method is already huge and begs refactoring. I'd say prior to this we should also refactor `publish` on its own. Only then we would be ready to go with this. And so I would indeed add another method than can publish in a batch.

Then we also could have an ability to Salt decide start "batching" after certain threshold without being explicitly configured to do so.

Reply:
If the implementation of this RFC requires changes to `publish`, I might agree that refactoring `publish` first could be a good idea.

If OTOH the implementation of this RFC uses `publish` as-is, I do not see a reason to withhold it until `publish` is refactored because of unrelated historical reasons. `publish` refactoring would be a separate task that could be equally be tackled before, after or even in parallel.

This is sometimes referred to as the "clean the kitchen, don't clean the garage" principle - is this acceptable?

Reply:
You are going to reuse publish's parts. Without prior refactoring (splitting to reusable methods etc) you will either doing it at the same time (and it is not a good idea) or duplicate code (not a good idea, again).

Sure, if you think that new publish method won't have any of the code duplication at all, go for it.

Reply:
Clarified in new commit, marking discussion resolved

Reply:
@cachedout are you OK with this?